modelforge/tests/test_sake.py

```diff
@@ -0,0 +1,227 @@
+    # Compare the rotated original dir_ij with the dir_ij from rotated positions
+    assert torch.allclose(rotated_reference_dir_ij, perturbed_dir_ij)
+
+    # Test that the interaction block is equivariant
+    sake_interaction = sake.interaction_modules[0]
+
+    reference_r = sake_interaction(
+        reference_prepared_input["atomic_embedding"],
+        reference_prepared_input["positions"],
+        reference_v,
+        reference_prepared_input["pair_indices"]
+    )
+
+    perturbed_r = sake_interaction(
+        perturbed_prepared_input["atomic_embedding"],
+        perturbed_prepared_input["positions"],
+        perturbed_v,
+        perturbed_prepared_input["pair_indices"],
+    )
+
+    perturbed_h, perturbed_x, perturbed_v = perturbed_r
+    reference_h, reference_x, reference_v = reference_r
+
+    # x and v are equivariant, h is invariant
+    assert torch.allclose(reference_h, perturbed_h)
+    assert torch.allclose(torch.matmul(reference_x, rotation_matrix), perturbed_x)
+    assert torch.allclose(torch.matmul(reference_v, rotation_matrix), perturbed_v)
+
+
+def test_spatial_attention_against_reference():
+    from modelforge.potential import CosineCutoff
+    from modelforge.potential import GaussianRBF
+    nr_atoms = 2
+    out_features = 11
+    hidden_features = 7
+    geometry_basis = 3
+    nr_heads = 5
+    nr_rbf = 50
+    cutoff_module = CosineCutoff(5.0 * unit.nanometer)
+    mf_sake_block = SAKEInteraction(
+        nr_atom_basis=out_features,
+        nr_edge_basis=hidden_features,
+        nr_edge_basis_hidden=hidden_features,
+        nr_atom_basis_hidden=hidden_features,
+        nr_atom_basis_post_norm_hidden=hidden_features,
+        nr_atom_basis_post_norm_out=hidden_features,
+        nr_atom_basis_velocity=hidden_features,
+        nr_coefficients=nr_heads * hidden_features,
+        nr_heads=nr_heads,
+        activation=torch.nn.SiLU(),
+        radial_basis_module=GaussianRBF(n_rbf=nr_rbf, cutoff=5.0 * unit.nanometer),
+        cutoff_module=cutoff_module,
+        epsilon=1e-5
+    )
+    ref_sake_interaction = reference_sake.layers.DenseSAKELayer(out_features=out_features,
+                                                                hidden_features=hidden_features,
+                                                                n_heads=nr_heads,
+                                                                cutoff=cutoff_module
+                                                                )
+
+    pairlist = torch.cartesian_prod(torch.arange(nr_atoms), torch.arange(nr_atoms))
+    nr_pairs = nr_atoms ** 2
+    pairlist = pairlist.T
+    idx_i, idx_j = pairlist
+
+    key = jax.random.PRNGKey(1884)
+    h_e_att = jax.random.normal(key, (nr_atoms, nr_atoms, hidden_features * nr_heads))
+    x_minus_xt = jax.random.normal(key, (nr_atoms, nr_atoms, geometry_basis))
+    x_minus_xt_norm = jnp.linalg.norm(x_minus_xt, axis=-1, keepdims=True)
```

Comment:
I think we can express this with pure numpy as well?

Reply:
ah yes sorry i was in the middle of converting this

Reply:
This is for the correctness tests (see above)